File: PocketFlicks/iPhone/Classes/WikipediaCache.m
// Copyright 2008 Cyrus Najmabadi
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "WikipediaCache.h"

#import "Application.h"
#import "FileUtilities.h"
#import "LinkedSet.h"
#import "Movie.h"
#import "NetworkUtilities.h"
#import "AppDelegate.h"
#import "Model.h"
#import "StringUtilities.h"
#import "ThreadingUtilities.h"
#import "Utilities.h"

@interface WikipediaCache()
@end


@implementation WikipediaCache

- (void) dealloc {
    [super dealloc];
}


- (id) initWithModel:(Model*) model_ {
    if (self = [super initWithModel:model_]) {
    }
    
    return self;
}


+ (WikipediaCache*) cacheWithModel:(Model*) model {
    return [[[WikipediaCache alloc] initWithModel:model] autorelease];
}


- (NSString*) wikipediaFile:(Movie*) movie {
    NSString* name = [[FileUtilities sanitizeFileName:movie.canonicalTitle] stringByAppendingPathExtension:@"plist"];
    return [[Application wikipediaDirectory] stringByAppendingPathComponent:name];
}


- (void) update:(NSArray*) movies {
    [self addPrimaryMovies:movies];
}


- (void) updateMovie:(Movie*) movie {
    [self addPrimaryMovie:movie];
}


- (void) updateMovieDetails:(Movie*) movie {
    NSString* path = [self wikipediaFile:movie];
    NSDate* lastLookupDate = [FileUtilities modificationDate:path];
    
    if (lastLookupDate != nil) {
        NSString* value = [FileUtilities readObject:path];
        if (value.length > 0) {
            // we have a real imdb value for this movie
            return;
        }
        
        // we have a sentinel.  only update if it's been long enough
        if (ABS(lastLookupDate.timeIntervalSinceNow) < (3 * ONE_DAY)) {
            return;
        }
    }
    
    NSString* url = [NSString stringWithFormat:@"http://%@.appspot.com/LookupWikipediaListings?q=%@", [Application host], [StringUtilities stringByAddingPercentEscapes:movie.canonicalTitle]];
    NSString* wikipediaAddress = [NetworkUtilities stringWithContentsOfAddress:url important:NO];
    if (wikipediaAddress == nil) {
        return;
    }
    
    // write down the response (even if it is empty).  An empty value will
    // ensure that we don't update this entry too often.
    [FileUtilities writeObject:wikipediaAddress toFile:path];
    if (wikipediaAddress.length > 0) {
        [AppDelegate minorRefresh];
    }
}


- (NSString*) wikipediaAddressForMovie:(Movie*) movie {
    return [FileUtilities readObject:[self wikipediaFile:movie]];
}

@end

File: trunk/NowPlaying/Android/src/org/metasyntactic/caches/websites/WikipediaCache.java
//Copyright 2008 Cyrus Najmabadi
//
//Licensed under the Apache License, Version 2.0 (the "License");
//you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.
package org.metasyntactic.caches.websites;

import java.io.File;
import java.util.Collections;
import java.util.List;

import org.metasyntactic.NowPlayingApplication;
import org.metasyntactic.NowPlayingModel;
import org.metasyntactic.caches.AbstractMovieCache;
import org.metasyntactic.data.Movie;
import org.metasyntactic.utilities.FileUtilities;
import org.metasyntactic.utilities.NetworkUtilities;
import org.metasyntactic.utilities.StringUtilities;

public class WikipediaCache extends AbstractMovieCache {

  public WikipediaCache(final NowPlayingModel model) {
    super(model);
  }

  private static String movieFileName(final Movie movie) {
    return FileUtilities.sanitizeFileName(movie.getCanonicalTitle());
  }

  private static File movieFilePath(final Movie movie) {
    return new File(NowPlayingApplication.wikipediaDirectory, movieFileName(movie));
  }

  public void update(final List<Movie> movies) {
    addPrimaryMovies(movies);
  }

  public void update(final Movie movie) {
    addPrimaryMovie(movie);
  }

  @Override protected void updateMovieDetails(final Movie movie) {
    if (movie == null) {
      return;
    }

    final File path = movieFilePath(movie);
    if (path.exists()) {
      final String address = FileUtilities.readString(path);
      if (address.length() > 0) {
        return;
      }

      if (FileUtilities.daysSinceNow(path) < 3) {
        return;
      }
    }

    final String url = "http://" + NowPlayingApplication.host + ".appspot.com/LookupWikipediaListings?q=" + StringUtilities
    .urlEncode(movie.getCanonicalTitle());

    final String imdbAddress = NetworkUtilities.downloadString(url, false);
    if (imdbAddress == null) {
      return;
    }

    FileUtilities.writeString(imdbAddress, path);
    NowPlayingApplication.refresh();
  }

  public static String getAddress(final Movie movie) {
    return FileUtilities.readString(movieFilePath(movie));
  }

  @Override
  protected List<File> getCacheDirectories() {
    return Collections.singletonList(NowPlayingApplication.wikipediaDirectory);
  }
}